client/beefy/src/import.rs

```diff
@@ -0,0 +1,211 @@
+			.validator_set(&block_id)
+			.map_err(|e| ConsensusError::ClientImport(e.to_string()))?
+			.ok_or_else(|| ConsensusError::ClientImport("Unknown validator set".to_string()))?;
+
+		decode_and_verify_commitment::<Block>(&encoded[..], number, &validator_set)
+	}
+
+	/// Import BEEFY justification: Send it to worker for processing and also append it to backend.
+	///
+	/// This function assumes:
+	/// - `justification` is verified and valid,
+	/// - the block referred by `justification` has been imported _and_ finalized.
+	fn import_beefy_justification_unchecked(
+		&self,
+		number: NumberFor<Block>,
+		justification: VersionedFinalityProof<NumberFor<Block>, Signature>,
+	) {
+		// Append the justification to the block in the backend.
+		if let Err(e) = self.backend.append_justification(
+			BlockId::Number(number),
+			(BEEFY_ENGINE_ID, justification.encode()),
+		) {
+			error!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, justification);
+		}
+		// Send the justification to the BEEFY voter for processing.
+		match justification {
+			// TODO #11838: Should not unpack, these channels should also use
+			// `VersionedFinalityProof`.
+			VersionedFinalityProof::V1(signed_commitment) => self
+				.justification_sender
+				.notify(|| Ok::<_, ()>(signed_commitment))
+				.expect("forwards closure result; the closure always returns Ok; qed."),
+		};
+	}
+}
+
+#[async_trait::async_trait]
+impl<Block, BE, Client, Runtime, I> BlockImport<Block>
+	for BeefyBlockImport<Block, BE, Client, Runtime, I>
+where
+	Block: BlockT,
+	BE: Backend<Block>,
+	Client: HeaderBackend<Block>,
+	I: BlockImport<
+			Block,
+			Error = ConsensusError,
+			Transaction = sp_api::TransactionFor<Runtime, Block>,
+		> + Send
+		+ Sync,
+	Runtime: ProvideRuntimeApi<Block> + Send + Sync,
+	Runtime::Api: BeefyApi<Block>,
+{
+	type Error = ConsensusError;
+	type Transaction = TransactionFor<Runtime, Block>;
+
+	async fn import_block(
+		&mut self,
+		mut block: BlockImportParams<Block, Self::Transaction>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+
+		let beefy_proof = block
+			.justifications
+			.as_mut()
+			.and_then(|just| {
+				let decoded = just
+					.get(BEEFY_ENGINE_ID)
+					.map(|encoded| self.decode_and_verify(encoded, number, hash));
+				// Remove BEEFY justification from the list before giving to `inner`;
+				// we will append it to backend ourselves at the end if all goes well.
+				just.remove(BEEFY_ENGINE_ID);
+				decoded
+			})
+			.transpose()
+			.unwrap_or(None);
+
+		// Run inner block import.
+		let inner_import_result = self.inner.import_block(block, new_cache).await?;
+
+		match (beefy_proof, &inner_import_result) {
+			(Some(proof), ImportResult::Imported(_)) => {
+				let status = self.client.info();
+				if number <= status.finalized_number &&
+					Some(hash) ==
+						self.client
```

Comment:
Is this `.hash()` call really required? Can we reuse `hash` from above here, or is it different? I wonder if following scenario may happen: (1) you have best finalized header = 100 and hash of finalized block 50 = `Hash50` (2) then we import alternate header 50 here with hash `AlternateHash50` and then when you call `client.hash(50)` here, it returns `Hash50`, so we'll try to append justification for `AlternateHash50` to `Hash50`.

Reply:
This check is to protect against exactly that:
`number` and `hash` are the block number and hash of the _currently importing block_ that came with current `justification`.

Right before actually importing `justification` we double-check that this block was in fact finalized (by GRANDPA or whatever finality gadget is underneath - we don't want to import any BEEFY justification for non-final blocks):
1. `number <= status.finalized_number` verifies that this block **could** be in the canonical fork/chain,
2. while `Some(hash) == self.client.hash(number)` builds on top of that and verifies _this_ `hash` is part of the _best_ fork (which is also finalized as checked in 1.)

> (1) you have best finalized header = 100 and hash of finalized block 50 = `Hash50` (2) then we import alternate header 50 here with hash `AlternateHash50` and then when you call `client.hash(50)` here, it returns `Hash50`, so we'll try to append justification for `AlternateHash50` to `Hash50`.

say we import both `AlternateHash50` and `Hash50`:
only `Hash50 == self.client.hash(50)` when `50 <= status.finalized_number`, so we drop any justification for `AlternateHash50` (**and log error - it can only happen if authorities are double voting**) and only import one from `Hash50`.

In theory this should not be possible (BEEFY justifications should only be possible for finalized blocks, never for alternates), but this check is here as an extra layer of defence.